be/src/olap/rowset/segment_v2/segment_iterator.cpp

```diff
@@ -310,34 +311,49 @@ Status SegmentIterator::next_batch(RowBlockV2* block) {
         block->set_num_rows(0);
         return Status::EndOfFile("no more data in segment");
     }
-    size_t rows_to_read = block->capacity();
-    while (rows_to_read > 0) {
-        if (_cur_rowid >= _row_ranges.get_range_to(_cur_range_id)) {
+    // initialize selection vector to all true
+    block->selection_vector()->set_all_true();
+
+    // check whether need to seek
+    if (_cur_rowid >= _row_ranges.get_range_to(_cur_range_id)) {
+        while (true) {
+            // step to next row range
+            ++_cur_range_id;
             // current row range is read over, trying to read from next range
             if (_cur_range_id >= _row_ranges.range_size() - 1) {
-                // there is no more row range
-                break;
+                block->set_num_rows(0);
+                return Status::EndOfFile("no more data in segment");
             }
-            // step to next row range
-            ++_cur_range_id;
-            _cur_rowid = _row_ranges.get_range_from(_cur_range_id);
             if (_row_ranges.get_range_count(_cur_range_id) == 0) {
                 // current row range is empty, just skip seek
                 continue;
             }
+            _cur_rowid = _row_ranges.get_range_from(_cur_range_id);
             for (auto cid : block->schema()->column_ids()) {
                 RETURN_IF_ERROR(_column_iterators[cid]->seek_to_ordinal(_cur_rowid));
             }
+            break;
         }
-        size_t to_read_in_range = std::min(rows_to_read, size_t(_row_ranges.get_range_to(_cur_range_id) - _cur_rowid));
-        RETURN_IF_ERROR(_next_batch(block, &to_read_in_range));
-        _cur_rowid += to_read_in_range;
-        rows_to_read -= to_read_in_range;
     }
-    block->set_num_rows(block->capacity() - rows_to_read);
+    // next_batch just return the rows in current row range
+    // it is easier to realize lazy materialization in the future
+    size_t rows_to_read = std::min(block->capacity(), size_t(_row_ranges.get_range_to(_cur_range_id) - _cur_rowid));
```

Comment:
Why don't you fill batch up?

Reply:
here just read a continuous row range because is it easier to do lazy materialization here in the future. and the caller can call next_batch to seek to the next continuous row range, the performance is equivalent.